### PR TITLE
rewrite canonicalize_syndication_url so it skips HEADing known silo URL formats

### PR DESCRIPTION
**Pull request by snarfed**
for #624, #628

@kylewm just fyi. no need to fully review this unless you're bored! (it's not
exactly clean commit by commit anyway; the full PR diff is probably simpler.)

i'm not entirely confident in it, but i'll probably try it out in prod for a bit
tomorrow and see what breaks. :P

### CODE REVIEW

File: original_post_discovery.py [hunk 1]
```diff
@@ -117,10 +117,10 @@ def resolve(urls):
...
     # different permalink formats to point to the same place (e.g.,
     # facebook user id instead of user name)
-    syndication_url = source.canonicalize_syndication_url(
-      util.follow_redirects(syndication_url).url)
-    originals.update(_posse_post_discovery(
-      source, activity, syndication_url, fetch_hfeed))
+    syndication_url = source.canonicalize_url(syndication_url)
+    if syndication_url:
+      originals.update(_posse_post_discovery(
+        source, activity, syndication_url, fetch_hfeed))
```
Comment (kylewm) [inline original_post_discovery.py]: oh interesting, so I was imagining it'd still follow redirects if the URL is unrecognized. This is more like: don't follow redirects (unless a service's implementation chooses to), we can always construct a canonical url out of the given url.

I'm fine with that change, just want to make sure we're on the same page.

**Comment by kylewm** [comment]
cool! this will make it a lot easier to change the permalink fetches to async... i'll be excited to try that soon.

File: original_post_discovery.py (hunk 1 above)
Comment (snarfed) [inline original_post_discovery.py]: oh no, it still follows redirects if the URL isn't known canonical. i just moved that into the new UrlCanonicalizer.

**Comment by snarfed** [comment]
huh, you think so? interesting, maybe!

File: original_post_discovery.py (hunk 1 above)
Comment (kylewm) [inline original_post_discovery.py]: ohh, I missed it because it's in webutil. ok got it.

**Comment by snarfed** [comment]
i deployed this at 6:57am pst, and it seemed to work ok, and didn't HEAD known silo post URLs, but it did start storing other silos' synd urls, eg it now has `SyndicatedPost`s for twitter, instagram, and G+ under my facebook account. https://console.cloud.google.com/datastore/entities/query?kind=SyndicatedPost&project=brid-gy&ns=&queryType=kind&filter=7%2F__key__|KEY|AN|27%2FKey%28FacebookPage,%20%27212038%27%29

hrmph. gotta fix that. in the meantime, i redeployed master at 7:11am pst.

also, wow, the coveralls PR check "fails" if coverage drops. intense!

**Comment by snarfed** [comment]
fixed the bug that was storing other silos' synd urls, tested locally, looks ok. deploying to prod now.

**Comment by snarfed** [comment]
looking good so far. no obvious errors, crawling a few test users worked fine, `SyndicatedPost`s look ok. https://console.cloud.google.com/datastore/entities/query?kind=SyndicatedPost&project=brid-gy&ns=&queryType=kind&sortDir=DESCENDING&sortCol=created

we'll see how latency looks after more data.

**Pull request merged**